src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp

```diff
@@ -72,6 +72,96 @@ mortars_apply_impl(const std::vector<std::array<size_t, Dim>>& initial_extents,
+                              evolution::dg::Tags::NormalCovector<Dim>>>>>*>
+        normal_covector_and_magnitude,
+    const gsl::not_null<MortarDataHistoryType*> mortar_data_history,
+    const Mesh<Dim>& new_mesh, const Element<Dim>& new_element,
+    const std::unordered_map<ElementId<Dim>, amr::Info<Dim>>& neighbor_info,
+    const std::pair<Mesh<Dim>, Element<Dim>>& old_mesh_and_element) {
+  const auto& [old_mesh, old_element] = old_mesh_and_element;
+  ASSERT(old_element.id() == new_element.id(),
+         "p-refinement should not have changed the element id");
+
+  std::unordered_set<DirectionalId<Dim>> old_mortar_ids;
+  for (const auto& [mortar_id, _] : *mortar_mesh) {
+    old_mortar_ids.emplace(mortar_id);
+  }
+
+  const bool mesh_changed = old_mesh != new_mesh;
+
+  for (const auto& [direction, neighbors] : new_element.neighbors()) {
+    const auto sliced_away_dimension = direction.dimension();
+    const auto old_face_mesh = old_mesh.slice_away(sliced_away_dimension);
+    const auto new_face_mesh = new_mesh.slice_away(sliced_away_dimension);
+    const bool face_mesh_changed = old_face_mesh != new_face_mesh;
+    if (face_mesh_changed) {
+      (*normal_covector_and_magnitude)[direction] = std::nullopt;
+    }
+    for (const auto& neighbor : neighbors) {
+      const DirectionalId<Dim> mortar_id{direction, neighbor};
+      if (old_mortar_ids.contains(mortar_id)) {
+        const auto new_neighbor_mesh = neighbors.orientation().inverse_map()(
+            neighbor_info.at(neighbor).new_mesh);
+        const auto& old_mortar_mesh = mortar_mesh->at(mortar_id);
+        auto new_mortar_mesh = ::dg::mortar_mesh(
+            new_mesh.slice_away(direction.dimension()),
+            new_neighbor_mesh.slice_away(direction.dimension()));
+        const bool mortar_mesh_changed = old_mortar_mesh != new_mortar_mesh;
+
+        if (mortar_mesh_changed or mesh_changed) {
+          // mortar_data does not need projecting as it has already been used
+          // and will be resized automatically
+          // mortar_size does not change as the mortar has not changed
+          // next_temporal_id does not change as the mortar has not changed
+          if (not mortar_data_history->empty()) {
+            auto& boundary_history = mortar_data_history->at(mortar_id);
+            auto local_history = boundary_history.local();
+            auto remote_history = boundary_history.remote();
+            const auto project_boundary_data =
+                [&new_mortar_mesh, &new_face_mesh, &new_mesh](
+                    const TimeStepId& /* id */,
+                    const gsl::not_null<::evolution::dg::MortarData<Dim>*>
+                        mortar_data) {
+                  p_project(mortar_data, new_mortar_mesh, new_face_mesh,
+                            new_mesh);
+                };
+            local_history.for_each(project_boundary_data);
+            remote_history.for_each(project_boundary_data);
```

Comment:
This passes the wrong face and volume meshes to `p_project` for the remote side, but I think either the associated remote fields are always unset or the values are never used.  Assuming that's true, add a comment noting that that's why it's OK.

Reply:
I added a new project function that only projects the mortar data to make this clear.  (I am adding the new function for the event-based refinement PR, so just seemed easier to add it now.)